forgejo: build from source

emilylange opened the pull request:
###### Description of changes

Build from source (frontend and backend) instead of relying on the release tarballs published with each release.
This does however add two additional hashes (`vendorHash` and `npmDepsHash`), making it a total of 3 hashes that need to be updated each time.

The old `updateScript` is no longer useful as of this change.
It was originally added because we weren't aware that gitea/forgejo/codeberg have "predictable URLs" for the previously needed release tarball.
Fixed in 0dd0b2103a25d2068462895faa852d535666a977

cc Gitea maintainers, in case they want this in `gitea` as well:
@disassembler @kolaente @Ma27 @techknowlogick

Suggested by @mweinelt in https://github.com/NixOS/nixpkgs/pull/224877#pullrequestreview-1373711014 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

urandom2 reviewed (approved):
i have not built or tested locally, but i am excited about the change, and everything looks great!

emilylange committed:
forgejo: build from source

instead of relying on the release tarballs published with each release.
This does however add two additional hashes (`vendorHash` and `npmDepsHash`), making it a total of 3 hashes that need to be updated each time.

The old `updateScript` is no longer useful as of this change.
It was originally added because we weren't aware that gitea/forgejo/codeberg have "predictable URLs" for the needed release tarball.
Fixed in 0dd0b2103a25d2068462895faa852d535666a977

IndeedNotJames commented:
I somehow forgot to VM test me latest iteration :(
So I switched back to an older version.
Also ran Forgejo in a [nixos-shell](https://github.com/Mic92/nixos-shell) and made sure everything looks and works as expected this time :)

Fix specific to Forgejo. Gitea could just use the previous version.

A little bit of background:
Forgejo has a custom `build/merge-forgejo-locales.go` script, that modifies and copies the translations from https://codeberg.org/forgejo/forgejo/src/tag/v1.19.1-0/options/locales to https://codeberg.org/forgejo/forgejo/src/tag/v1.19.1-0/options/locale

Technically, the Makefile also has a `go generate` section, but I couldn't really find out if this is actually needed in our case.
That section also has the following comment:

> We generate the backend before the frontend in case we in future we want to generate things in the frontend from generated files in backend

https://codeberg.org/forgejo/forgejo/src/tag/v1.19.1-0/Makefile#L764-L774

This has the potential to make it a cyclic dependency between backend and frontend due to [`static-root-path.patch`](https://github.com/NixOS/nixpkgs/blob/149c5e55062cdb6ca5ef7ad18a660e26db6b36ca/pkgs/applications/version-management/gitea/static-root-path.patch), in a future update :/
But this isn't the case yet. I guess only time will tell? :sweat_smile:

emilylange committed:
nixosTests.gitea: fix sshd race condition

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2108

SuperSandro2000 commented on pkgs/applications/version-management/forgejo/package-json-npm-build-frontend.patch:
```diff
@@ -0,0 +1,14 @@
+-  ]
++  ],
++  "scripts": {
++    "build": "npx webpack"
```
npx potentially tries to download packages, we want to run the webpack from our node_modules directory.

SuperSandro2000 commented on pkgs/applications/version-management/forgejo/default.nix:
```diff
@@ -59,15 +76,25 @@ buildGoModule rec {
       --prefix PATH : ${lib.makeBinPath [ bash git gzip openssh ]}
   '';
 
+  ## $data is not available in go-modules.drv and preBuild isn't needed
```
```suggestion
  # $data is not available in go-modules.drv and preBuild isn't needed
```

even if it isn't required, if it does nothing undesirable we should just leave it as is.

IndeedNotJames replied:
yeah, double `#` is a bit of a habit of mine :sweat_smile: 

will remove it :+1:
